rs_lib/src/file_fetcher/mod.rs

```diff
@@ -0,0 +1,1038 @@
+        SendResponse::Redirect(headers) => {
+          let new_url = resolve_redirect_from_headers(args.url, &headers)
+            .map_err(|err| {
+              FetchNoFollowErrorKind::RedirectHeaderParse(*err).into_box()
+            })?;
+          Ok(SendRequestResponse::Redirect(
+            new_url,
+            response_headers_to_headers_map(headers),
+          ))
+        }
+        SendResponse::Success(headers, body) => Ok(SendRequestResponse::Code(
+          body,
+          response_headers_to_headers_map(headers),
+        )),
+      },
+      Err(err) => match err {
+        SendError::Failed(err) => Err(
+          FetchNoFollowErrorKind::FetchingRemote {
+            specifier: args.url.clone(),
+            source: err,
+          }
+          .into_box(),
+        ),
+        SendError::NotFound => {
+          Err(FetchNoFollowErrorKind::NotFound(args.url.clone()).into_box())
+        }
+        SendError::StatusCode(status_code) => Err(
+          FetchNoFollowErrorKind::ClientError {
+            specifier: args.url.clone(),
+            status_code,
+          }
+          .into_box(),
+        ),
+      },
+    }
+  }
+
+  /// Fetch a source file from the local file system.
+  pub fn fetch_local(
+    &self,
+    specifier: &Url,
+  ) -> Result<Option<File>, FetchLocalError> {
+    let local = url_to_file_path(specifier)?;
+    // If it doesnt have a extension, we want to treat it as typescript by default
+    let headers = if local.extension().is_none() {
+      Some(HashMap::from([(
+        "content-type".to_string(),
+        "application/typescript".to_string(),
+      )]))
+    } else {
+      None
+    };
+    let bytes = match self.env.read_file_bytes(&local) {
+      Ok(bytes) => bytes,
+      Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
+        return Ok(None);
+      }
+      Err(err) => {
+        return Err(
+          FetchLocalErrorKind::ReadingFile(FailedReadingLocalFileError {
+            specifier: specifier.clone(),
+            source: err,
+          })
+          .into_box(),
+        );
+      }
+    };
+
+    Ok(Some(File {
+      specifier: specifier.clone(),
+      maybe_headers: headers,
+      source: bytes.into(),
+    }))
+  }
+}
+
+fn response_headers_to_headers_map(response_headers: HeaderMap) -> HeadersMap {
+  let mut result_headers = HashMap::with_capacity(response_headers.len());
+  // todo(dsherret): change to consume to avoid allocations
+  for key in response_headers.keys() {
+    let key_str = key.to_string();
+    let values = response_headers.get_all(key);
+    // todo(dsherret): this seems very strange storing them comma separated
+    // like this... what happens if a value contains a comma?
+    let values_str = values
+      .iter()
+      .filter_map(|e| Some(e.to_str().ok()?.to_string()))
+      .collect::<Vec<String>>()
+      .join(",");
+    result_headers.insert(key_str, values_str);
+  }
+  result_headers
+}
```

Comment:
This function is strange, but it's mostly a move down from the CLI. I added some todos.